packages/react-ui/src/app/features/builder/hooks/use-paste.ts

```diff
@@ -5,17 +5,61 @@ import {
 } from '@openops/components/ui';
 import {
   Action,
+  ActionType,
   flowHelper,
   FlowOperationType,
   FlowVersion,
   isNil,
+  openOpsId,
   StepLocationRelativeToParent,
 } from '@openops/shared';
 import { useCallback } from 'react';
 import { useBuilderStateContext } from '../builder-hooks';
 import { useApplyOperationAndPushToHistory } from '../flow-version-undo-redo/hooks/apply-operation-and-push-to-history';
 import { useCenterWorkflowViewOntoStep } from './center-workflow-view-onto-step';
 
+export const assignNewIdsToActions = (action: Action): Action => {
+  const newAction = { ...action };
+  newAction.id = openOpsId();
+
+  if (newAction.nextAction) {
+    newAction.nextAction = assignNewIdsToActions(newAction.nextAction);
+  }
+
+  if (newAction.type === ActionType.BRANCH) {
+    if (newAction.onSuccessAction) {
+      newAction.onSuccessAction = assignNewIdsToActions(
+        newAction.onSuccessAction,
+      );
+    }
+    if (newAction.onFailureAction) {
```

Comment:
Idea: we already have code that generates slugs.
Let's put the IDs only there, this way we know it's called by all relevant places.